File: dong_chan/Python/알고리즘_자료구조/트리/트리순회.py
# 9월7일 백준 트리 순회 실버1

import sys


class Node():
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.left = right


class Tree():
    def __init__(self):
        self.root = None

# input 값을 트리에 저장할 함수 정의

# 루트 노드는 항상 A


def TreeInput(value, left, right):
    pass


# 정답 확인 = 딕셔너리를 이용한 트리

N = int(sys.stdin.readline().strip())
tree = {}

for n in range(N):
    root, left, right = sys.stdin.readline().strip().split()
    tree[root] = [left, right]


def preorder(root):
    if root != '.':
        print(root, end='')  # root
        preorder(tree[root][0])  # left
        preorder(tree[root][1])  # right


def inorder(root):
    if root != '.':
        inorder(tree[root][0])  # left
        print(root, end='')  # root
        inorder(tree[root][1])  # right


def postorder(root):
    if root != '.':
        postorder(tree[root][0])  # left
        postorder(tree[root][1])  # right
        print(root, end='')  # root


preorder('A')
print()
inorder('A')
print()
postorder('A')

# 문제 이해
# input 값의 타입은 무엇인가 ?
# n 의 범위
# output

File: JavaScript/3주차/트리/트리_순회.js
let fs = require('fs');
let input = fs.readFileSync('io.txt').toString().trim().split('\n');

let n = parseInt(input.shift());
let arr = input.map((a) => a.split(' '));

class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

class BT {
  constructor() {
    this.root = null;
    this.nodes = {}; // 노드 저장용 객체
  }

  insert(value, left, right) {
    if (!this.nodes[value]) this.nodes[value] = new Node(value);
    if (left !== '.') this.nodes[left] = new Node(left);
    if (right !== '.') this.nodes[right] = new Node(right);
    if (left !== '.') this.nodes[value].left = this.nodes[left];
    if (right !== '.') this.nodes[value].right = this.nodes[right];

    if (!this.root) this.root = this.nodes['A']; // 루트 설정
  }

  // 📌 전위 순회 (Preorder)
  preOrder() {
    this._preorderRec(this.root);
    console.log();
  }

  _preorderRec(node) {
    if (!node) return;
    process.stdout.write(node.value);
    this._preorderRec(node.left);
    this._preorderRec(node.right);
  }

  // 📌 중위 순회 (Inorder)
  inOrder() {
    this._inorderRec(this.root);
    console.log();
  }

  _inorderRec(node) {
    if (!node) return;
    this._inorderRec(node.left);
    process.stdout.write(node.value);
    this._inorderRec(node.right);
  }

  // 📌 후위 순회 (Postorder)
  postOrder() {
    this._postorderRec(this.root);
    console.log();
  }

  _postorderRec(node) {
    if (!node) return;
    this._postorderRec(node.left);
    this._postorderRec(node.right);
    process.stdout.write(node.value);
  }
}

let bt = new BT();

for (let i of arr) {
  // console.log(i);
  bt.insert(i[0], i[1], i[2]);
}

bt.preOrder();
bt.inOrder();
bt.postOrder();
